paperqa/agents/main.py

```diff
@@ -0,0 +1,383 @@
+    SimpleProfiler,
+)
+from .search import SearchDocumentStorage, SearchIndex
+from .tools import (
+    EmptyDocsError,
+    GatherEvidenceTool,
+    GenerateAnswerTool,
+    PaperSearchTool,
+    SharedToolState,
+    query_to_tools,
+    status,
+)
+
+logger = logging.getLogger(__name__)
+agent_logger = logging.getLogger("agent_callers")
+
+
+def configure_agent_logging(verbosity: int = 0, default_level: int = logging.INFO):
+    """Default to INFO level, but suppress loquacious loggers."""
+    verbosity_map = {
+        0: {
+            "paperqa.agents.docs": logging.WARNING,
+            "paperqa.agents.main": logging.WARNING,
+            "anthropic": logging.WARNING,
+            "openai": logging.WARNING,
+            "httpx": logging.WARNING,
+            "agent_callers": logging.INFO,
+            "paperqa.agents.models": logging.WARNING,
+            "paperqa.agents.search": logging.INFO,
+        },
+        1: {
+            "paperqa.agents.docs": logging.INFO,
+            "paperqa.agents.main": logging.WARNING,
+            "anthropic": logging.WARNING,
+            "openai": logging.WARNING,
+            "httpx": logging.WARNING,
+            "agent_callers": logging.INFO,
+            "paperqa.models": logging.INFO,
+            "paperqa.agents.search": logging.INFO,
+        },
+        2: {
+            "paperqa.agents.docs": logging.DEBUG,
+            "paperqa.agents.main": logging.DEBUG,
+            "anthropic": logging.INFO,
+            "httpx": logging.INFO,
+            "openai": logging.INFO,
+            "agent_callers": logging.DEBUG,
+            "paperqa.models": logging.DEBUG,
+            "paperqa.agents.search": logging.DEBUG,
+        },
+    }
+    for logger_name in logging.Logger.manager.loggerDict:
+        logger = logging.getLogger(logger_name)
+        logger.setLevel(
+            verbosity_map.get(min(verbosity, 2), {}).get(logger_name, default_level)
+        )
+
+
+async def aagent_query(
+    query: str | QueryRequest,
+    docs: Docs | None = None,
+    agent_type: str = "OpenAIFunctionsAgent",
+    verbosity: int = 0,
+    index_directory: str | os.PathLike | None = None,
+) -> AnswerResponse:
+
+    configure_agent_logging(verbosity=verbosity)
+
+    if isinstance(query, str):
+        query = QueryRequest(query=query)
+
+    if docs is None:
+        docs = Docs()
+
+    if index_directory is None:
+        index_directory = pqa_directory("indexes")
+
+    # in-place modification of the docs object to match query
+    update_doc_models(
+        docs,
+        query,
+    )
+
+    search_index = SearchIndex(
+        fields=["file_location", "body", "question"],
+        index_name="answers",
+        index_directory=index_directory,
+        storage=SearchDocumentStorage.JSON_MODEL_DUMP,
+    )
+
+    response = await run_agent(docs, query, agent_type)
+
+    agent_logger.debug(f"agent_response: {response}")
+    truncation_max = 1_000_000 if verbosity > 1 else 1500 * (verbosity + 1)
```

Comment:
What are the units here? Wdyt of naming `truncation_chars`? I think "max" is implied with truncation

Reply:
characters